### PR TITLE
[9.0] Add web_access_rule_buttons

### PR DESCRIPTION
**Pull request by guewen**
Disable buttons which lead to action disallowed by record rules.

### CODE REVIEW

**Comment by yvaucher** [comment]
:+1:

File: web_access_rule_buttons/README.rst [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/162/9.0
+
+Known issues / Roadmap
+======================
+
+* Additional requests will be issued when a record is loaded in a form view in
+  order to check if the user has the accesses.
```
Comment (elicoidal) [inline web_access_rule_buttons/README.rst]: s/accesses/access right

**Comment by yvaucher** [comment]
When trying to create an event from calendar, I get this error:

```
Uncaught TypeError: Cannot read property 'find' of undefined
http://localhost:8501/web_access_rule_buttons/static/src/js/web_access_rule_buttons.js:38
Traceback:
TypeError: Cannot read property 'find' of undefined
    at OdooClass.FormView.include.show_hide_edit_button (http://localhost:8501/web_access_rule_buttons/static/src/js/web_access_rule_buttons.js:38:33)
at Object.<anonymous> (http://localhost:8501/web_access_rule_buttons/static/src/js/web_access_rule_buttons.js:33:16)
at Object.<anonymous> (http://localhost:8501/web/static/lib/jquery/jquery.js:3276:89)
at fire (http://localhost:8501/web/static/lib/jquery/jquery.js:3119:58)
at Object.self.fireWith [as resolveWith] (http://localhost:8501/web/static/lib/jquery/jquery.js:3231:49)
at Object.deferred.(anonymous function) (http://localhost:8501/web/static/lib/jquery/jquery.js:3321:62)
at fire (http://localhost:8501/web/static/lib/jquery/jquery.js:3119:58)
at Object.self.fireWith [as resolveWith] (http://localhost:8501/web/static/lib/jquery/jquery.js:3231:49)
at Object.<anonymous> (http://localhost:8501/web/static/lib/jquery/jquery.js:3283:104)
at fire (http://localhost:8501/web/static/lib/jquery/jquery.js:3119:58)
```

File: web_access_rule_buttons/static/src/js/web_access_rule_buttons.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      this.dataset.call('check_access_rule_all',
+                        [[this.datarecord.id],
+                         ['write']])
+        .then(function(accesses) {
+          self.show_hide_edit_button(accesses.write);
+        });
+    },
+
+    show_hide_edit_button: function(access) {
+      var button = this.$buttons.find('.oe_form_button_edit');
```
Comment (yvaucher) [inline web_access_rule_buttons/static/src/js/web_access_rule_buttons.js]: In a popup window there are no buttons thus it is needed to ensure buttons are defined first

Comment (yvaucher) [inline web_access_rule_buttons/static/src/js/web_access_rule_buttons.js]: I fixed it here:

https://github.com/guewen/web/pull/1

**Comment by guewen** [comment]
@yvaucher I merged your correction, thanks
@elicoidal fixed, thanks for your proofreading

File: web_access_rule_buttons/static/src/js/web_access_rule_buttons.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  "use strict";
+
+  var core = require('web.core');
+
+  var FormView = require('web.FormView');
+  var ListView = require('web.ListView');
+
+  var QWeb = core.qweb;
+
+  var _t = core._t;
```
Comment (petrus-v) [inline web_access_rule_buttons/static/src/js/web_access_rule_buttons.js]: I know we are a lot to leave those 2 lines in the beginning of the class but I still wondering why we are setting them if we are not using them?

Comment (legalsylvain) [inline web_access_rule_buttons/static/src/js/web_access_rule_buttons.js]: indeed, it's like unused python librairies import.
+1 to remove those 2 lines.

File: web_access_rule_buttons/static/src/js/web_access_rule_buttons.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        .then(function(accesses) {
+          self.show_hide_edit_button(accesses.write);
+        });
+    },
+
+    show_hide_edit_button: function(access) {
+      if (this.$buttons) {
+          var button = this.$buttons.find('.oe_form_button_edit');
+          if(access) {
+            button.removeAttr('disabled');
```
Comment (petrus-v) [inline web_access_rule_buttons/static/src/js/web_access_rule_buttons.js]: Is there a way that button will be undefined (if some module play to remove '.oe_form_button_edit' ?)?

Comment (petrus-v) [inline web_access_rule_buttons/static/src/js/web_access_rule_buttons.js]: According [.prop()](http://api.jquery.com/prop/) jQuery documentation I would probably prefer the following syntax

```
button.prop('disabled', access);
```

Comment (guewen) [inline web_access_rule_buttons/static/src/js/web_access_rule_buttons.js]: Thanks for the hint!
Done. (but with `!access` as we _activate_ `disabled` when we have no access and _disable_ it when we have access)

File: web_access_rule_buttons/static/src/js/web_access_rule_buttons.js (hunk 3 above)
Comment (guewen) [inline web_access_rule_buttons/static/src/js/web_access_rule_buttons.js]: I agree, they are removed now.

**Comment by guewen** [comment]
Thanks for your suggestions, I improved my proposal by applying them.

**Comment by petrus-v** [comment]
Thanks!

:+1: Review only

**Comment by bealdav** [comment]
Thanks Guewen. Any chance to see that in odoo master ?

**Pull request merged**